### PR TITLE
Fix janks and memory leaks in `CupertinoPageTransition` and `CupertinoFullscreenDialogTransition` [prod-leak-fix]

### PR DESCRIPTION
**Pull request by ValentinVignal**
Initially, part of https://github.com/flutter/flutter/issues/141198 to fix leak memories in `CupertinoPageTransition`. 
But it was overlapping with https://github.com/flutter/flutter/pull/147133 which got merged first.
During my work, I noticed the `CupertinoPageTransition` and `CupertinoFullscreenDialogTransition`'s animations states were lost when a push was interrupted with a pop. When that is the case, it should use the forward curve for the reverse animation to avoid discontinuities. This PR fixes it

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by ValentinVignal** [comment]
cc @polina-c

File: packages/flutter/lib/src/cupertino/route.dart [hunk 1]
```diff
@@ -387,48 +387,97 @@ class CupertinoPageTransition extends StatelessWidget {
...
+  CurvedAnimation? _backingPrimaryPositionAnimation;
+  // When this page is coming in to cover another page.
+  Animation<Offset>  get _primaryPositionAnimation => (_backingPrimaryPositionAnimation ?? widget.primaryRouteAnimation).drive(_kRightMiddleTween);
+
+  CurvedAnimation? _backingSecondaryPositionAnimation;
+  // When this page is coming in to cover another page.
+  Animation<Offset>  get _secondaryPositionAnimation => (_backingSecondaryPositionAnimation ?? widget.secondaryRouteAnimation).drive(_kMiddleLeftTween);
+
+  CurvedAnimation? _backingPrimaryShadowAnimation;
+  // When this page is coming in to cover another page.
```
Comment (polina-c) [inline packages/flutter/lib/src/cupertino/route.dart:428]: _primaryShadowAnimation did not have this comment in original code

File: packages/flutter/lib/src/cupertino/route.dart [hunk 2]
```diff
@@ -387,48 +387,97 @@ class CupertinoPageTransition extends StatelessWidget {
...
 
+  @override
+  State<CupertinoPageTransition> createState() => _CupertinoPageTransitionState();
+}
+
+class _CupertinoPageTransitionState extends State<CupertinoPageTransition> {
+
+  CurvedAnimation? _backingPrimaryPositionAnimation;
+  // When this page is coming in to cover another page.
+  Animation<Offset>  get _primaryPositionAnimation => (_backingPrimaryPositionAnimation ?? widget.primaryRouteAnimation).drive(_kRightMiddleTween);
```
Comment (polina-c) [inline packages/flutter/lib/src/cupertino/route.dart:421]: // -> ///

Here and in other comment.

File: packages/flutter/lib/src/cupertino/route.dart [hunk 3]
```diff
@@ -387,48 +387,97 @@ class CupertinoPageTransition extends StatelessWidget {
...
+
+  CurvedAnimation? _backingSecondaryPositionAnimation;
+  // When this page is coming in to cover another page.
+  Animation<Offset>  get _secondaryPositionAnimation => (_backingSecondaryPositionAnimation ?? widget.secondaryRouteAnimation).drive(_kMiddleLeftTween);
```
Comment (polina-c) [inline packages/flutter/lib/src/cupertino/route.dart:425]: Do we really want to invoke method `drive` on every invocation of the getter?

Comment (polina-c) [reply packages/flutter/lib/src/cupertino/route.dart:425]: May be _setBackingCurveAnimations is right place to invoke it?

File: packages/flutter/lib/src/cupertino/route.dart [hunk 4]
```diff
@@ -387,48 +387,106 @@ class CupertinoPageTransition extends StatelessWidget {
...
+  @override
+  void didUpdateWidget(CupertinoPageTransition oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (
+      oldWidget.linearTransition != widget.linearTransition ||
+      oldWidget.primaryRouteAnimation != widget.primaryRouteAnimation ||
+      oldWidget.secondaryRouteAnimation != widget.secondaryRouteAnimation) {
+      _setAnimations();
+    }
+  }
```
Comment (ValentinVignal) [inline packages/flutter/lib/src/cupertino/route.dart:447]: @polina-c I'm having some trouble with some tests. (ex: `packages/flutter/test/cupertino/route_test.dart` "Animated push/pop is not linear").

https://github.com/flutter/flutter/blob/98685a099fdde0b0424b5665debbe508a5c697e5/packages/flutter/test/cupertino/route_test.dart#L928

Looking at the values (the `.dx`s) of the test, it looks like they are correct.

However, with my changes, it fails here:

https://github.com/flutter/flutter/blob/98685a099fdde0b0424b5665debbe508a5c697e5/packages/flutter/test/cupertino/route_test.dart#L965-L968

```console
The following TestFailure was thrown running a test:
Expected: -197.0 (±1.0)
  Actual: <-265.3022323052088>
   Which: -265.3022323052088 is not in the range of -197.0 (±1.0).
```

It looks like, with my changes, it is not using the `reverseCurve` given to the `CurvedAnimation` anymore.

Interestingly, the test passes if I do 

```suggestion
  @override
  void didUpdateWidget(CupertinoPageTransition oldWidget) {
    super.didUpdateWidget(oldWidget);
    _setAnimations();
  }
```

Do you have an idea of what could be going on?

File: packages/flutter/test/cupertino/route_test.dart [hunk 5]
```diff
@@ -960,8 +974,14 @@ void main() {
 
     // Finish the rest of the animation
     await tester.pump(const Duration(milliseconds: 350));
+    // Give time to the animation to finish and update its status to
+    // AnimationState.completed, so the reverse curved can be used in the next
+    // step.
+    await tester.pump(const Duration(milliseconds: 1));
```
Comment (ValentinVignal) [inline packages/flutter/test/cupertino/route_test.dart:980]: The issue was that the parent of the curved animation needs to change its status from `forward` to `completed` so it doesn't ignore when it the parents updates it to `reverse`

https://github.com/flutter/flutter/blob/1a905d508de1608ba68ee1bd4492e8cb3a428709/packages/flutter/lib/src/animation/animations.dart#L427-L431

This is how I managed to do it, I don't know if there is a better way to do that?

`tester.pump()` alone doesn't work. `tester.pumpAndSettle()` works but I believe it is because it has a default duration of 100ms.

Would you have a better approach?

File: packages/flutter/test/cupertino/route_test.dart [hunk 6]
```diff
@@ -784,47 +784,56 @@ void main() {
...
 
-    await tester.pump(const Duration(milliseconds: 40));
-    expect(tester.getTopLeft(find.byType(Placeholder)).dx, moreOrLessEquals(-111.0, epsilon: 1.0));
+    await tester.pump(const Duration(milliseconds: 50));
+    expect(tester.getTopLeft(find.byType(Placeholder)).dx, moreOrLessEquals(-69.0, epsilon: 1.0));
 
-    await tester.pump(const Duration(milliseconds: 40));
-    expect(tester.getTopLeft(find.byType(Placeholder)).dx, moreOrLessEquals(-161.0, epsilon: 1.0));
+    await tester.pump(const Duration(milliseconds: 50));
+    expect(tester.getTopLeft(find.byType(Placeholder)).dx, moreOrLessEquals(-136.0, epsilon: 1.0));
```
Comment (ValentinVignal) [inline packages/flutter/test/cupertino/route_test.dart:797]: I believe this test was outdated, the total duration of the page is 500ms, so it makes more sense to update with 10 chunks of 50ms instead of 40ms.

During the forward (push), the value _"didn't change"_ (meaning the test was passing). I only updated them because I changed the duration in the `pump()`s.

File: packages/flutter/test/cupertino/route_test.dart [hunk 7]
```diff
@@ -784,47 +784,56 @@ void main() {
...
     await tester.pump();
 
-    await tester.pump(const Duration(milliseconds: 40));
-    expect(tester.getTopLeft(find.byType(Placeholder)).dx, moreOrLessEquals(-83.0, epsilon: 1.0));
+    await tester.pump(const Duration(milliseconds: 50));
+    expect(tester.getTopLeft(find.byType(Placeholder)).dx, moreOrLessEquals(-197.0, epsilon: 1.0));
 
-    await tester.pump(const Duration(milliseconds: 360));
+    await tester.pump(const Duration(milliseconds: 450));
     expect(tester.getTopLeft(find.byType(Placeholder)).dx, moreOrLessEquals(-0.0, epsilon: 1.0));
```
Comment (ValentinVignal) [inline packages/flutter/test/cupertino/route_test.dart:837]: Here the values did changes (meaning the test was falling), but from the code, it looks like they were wrong:

The reverse animation is `Curves.easeInToLinear` and the `Placeholder` goes from `-800 / 3 = -266.67 px` to `0px` (the screen is 800px wide and `_kMiddleLeftTween` only uses 1/3 rd of the space).

The duration of the animation is 500ms:

`- (800 / 3) * Curves.easeInToLinear.transform(50 / 500)` is around 197

**Comment by ValentinVignal** [comment]
@polina-c I managed to "_fix"_ the tests in `route_test.dart`, I tried to explain them, tell me if something is obscure

File: packages/flutter/test/cupertino/route_test.dart [hunk 8]
```diff
@@ -784,47 +784,56 @@ void main() {
...
+
+
     // Exit animation
     await tester.tap(find.text('Close'));
     await tester.pump();
 
-    await tester.pump(const Duration(milliseconds: 40));
-    expect(tester.getTopLeft(find.byType(Placeholder)).dx, moreOrLessEquals(-83.0, epsilon: 1.0));
+    await tester.pump(const Duration(milliseconds: 50));
+    expect(tester.getTopLeft(find.byType(Placeholder)).dx, moreOrLessEquals(-197.0, epsilon: 1.0));
```
Comment (polina-c) [inline packages/flutter/test/cupertino/route_test.dart:834]: This change from -83 to -197 looks alerting to me. 
The test passes at flutter master with -83. And it stops passing with -83 after your change. That means your change is not just improvement in memory management, it modified functionality, that should not happen.

Comment (ValentinVignal) [reply packages/flutter/test/cupertino/route_test.dart:834]: I agree, I was trying to _"fix"_ those changes in the tests. However, with my changes the behavior changes and I will have to update those values.

Before, a new `CurvedAnimation` was recreated at every build (in the constructor of the stateless widget `CupertinoPageTransition`). Now, it is reusing the same one. This leads to a change in the functionality (which I believe is correct).

As a reminder, here are some useful values:

```dart
Curves.easeInToLinear.transform(0.72); // 0.3149519026279449
- (800 / 3) * Curves.easeInToLinear.transform(0.72);  // -83.98717403411865
Curves.linearToEaseOut.transform(0.72); // 0.979008914809674
- (800 / 3) * Curves.linearToEaseOut.transform(0.72); // -261.0690439492464
```

And the secondary animation is 

```dart
_secondPositionCurvedAnimation = CurvedAnimation(
  parent: widget.secondaryRouteAnimation,
  curve: Curves.linearToEaseOut,
  reverseCurve: Curves.easeInToLinear,
);
```


Before, the test was animating the push of the page from 0ms to 400ms (the entire duration of the push animation is 500ms) and then popping. So the push animation was not completed when being reversed. During the forward animation, the curve used is `Curves.linearToEaseOut` (which is correct).
During the pop/reverse animation, because the `CurvedAnimation` is recreated, it was using the `reversedAnimation: Curves.easeInToLinear` which led to this `-83` value.


But this shouldn't happen, if the animation is not complete, it should use its forward curve during the reverse animation. I guess this is to prevent jumps and janks.

This is what those lines are doing:

https://github.com/flutter/flutter/blob/89a4ffaad5db20e2b35754a665902a8c78bbca09/packages/flutter/lib/src/animation/animations.dart#L427-L432

https://github.com/flutter/flutter/blob/89a4ffaad5db20e2b35754a665902a8c78bbca09/packages/flutter/lib/src/animation/animations.dart#L434-L436

https://github.com/flutter/flutter/blob/89a4ffaad5db20e2b35754a665902a8c78bbca09/packages/flutter/lib/src/animation/animations.dart#L450-L457


Now that the `CurvedAnimation` is stored in the state and reused, it keeps using the forward curve and doesn't _"jump"_ to the `reversedCurve`, which I believe fixes a jank/jump that currently exists on the master channel.

---

For the test, I decided the modify it so the push animation finishes and the pop uses the `reversedCurved` by changing the durations from 40ms to 50ms. But I can revert those duration changes and make the pop use the forward curve instead.

**Comment by ValentinVignal** [comment]
@polina-c This was the same as https://github.com/flutter/flutter/pull/147133 but I believe there are some flows in the PR that this one fixes. I think we can keep it open

**Comment by polina-c** [comment]
@Piinks , Valentin is suggesting functional fix in the animation, as side effect of the leak fix. Can you help to review it or suggest someone who has enough expertise?

**Comment by polina-c** [comment]
> @polina-c This was the same as #147133 but I believe there are some flows in the PR that this one fixes. I think we can keep it open

Nope, if the change is good, we want to push it forward. You believe it is good, right?
It just will take more time to explain it to others :)

**Comment by ValentinVignal** [comment]
> You believe it is good, right?

I've at least started to convince myself it is good haha 😅 It for sure needs a confirmation from someone with expertise

File: packages/flutter/lib/src/cupertino/route.dart [hunk 9]
```diff
@@ -389,97 +395,96 @@ class CupertinoPageTransition extends StatefulWidget {
     required this.linearTransition,
   });
 
-  /// The widget below this widget in the tree.
-  final Widget child;
-
-  ///  * `primaryRouteAnimation` is a linear route animation from 0.0 to 1.0
-  ///    when this screen is being pushed.
+  /// A linear route animation from 0.0 to 1.0 when this screen is being pushed.
```
Comment (Piinks) [inline packages/flutter/lib/src/cupertino/route.dart:398]: Is this only linear based on `linearTransition`?

File: packages/flutter/lib/src/cupertino/route.dart [hunk 10]
```diff
@@ -389,97 +395,96 @@ class CupertinoPageTransition extends StatefulWidget {
...
   final Animation<double> primaryRouteAnimation;
 
-  ///  * `secondaryRouteAnimation` is a linear route animation from 0.0 to 1.0
-  ///    when another screen is being pushed on top of this one.
+  /// A linear route animation from 0.0 to 1.0 when another screen is being
```
Comment (Piinks) [inline packages/flutter/lib/src/cupertino/route.dart:401]: Same here, is this only linear based on `linearTransition`?

File: packages/flutter/lib/src/cupertino/route.dart [hunk 11]
```diff
@@ -389,97 +395,96 @@ class CupertinoPageTransition extends StatefulWidget {
...
+  CurvedAnimation? _primaryPositionCurvedAnimation;
 
-  // When this page is coming in to cover another page.
+  /// When this page is coming in to cover another page.
   late Animation<Offset> _primaryPositionAnimation;
-  // When this page is becoming covered by another page.
+
+  CurvedAnimation? _secondaryPositionCurvedAnimation;
+
+  /// When this page is becoming covered by another page.
```
Comment (Piinks) [inline packages/flutter/lib/src/cupertino/route.dart:426]: Since these are private, the `///` public docs do not need to be added. For the others, can you add the comments back?

File: packages/flutter/lib/src/cupertino/route.dart [hunk 12]
```diff
@@ -389,97 +395,96 @@ class CupertinoPageTransition extends StatefulWidget {
...
-  // Curve of primary page which is coming in to cover another page.
-  CurvedAnimation? _primaryPositionCurve;
-  // Curve of secondary page which is becoming covered by another page.
-  CurvedAnimation? _secondaryPositionCurve;
-  // Curve of primary page's shadow.
-  CurvedAnimation? _primaryShadowCurve;
 
   @override
-  void initState() {
+  void initState()  {
```
Comment (Piinks) [inline packages/flutter/lib/src/cupertino/route.dart:433]:
```suggestion
  void initState() {
```

File: packages/flutter/lib/src/cupertino/route.dart [hunk 13]
```diff
@@ -389,97 +395,96 @@ class CupertinoPageTransition extends StatefulWidget {
...
   @override
-  void didUpdateWidget(covariant CupertinoPageTransition oldWidget) {
+  void didUpdateWidget(CupertinoPageTransition oldWidget) {
     super.didUpdateWidget(oldWidget);
-    if (oldWidget.primaryRouteAnimation != widget.primaryRouteAnimation
-    || oldWidget.secondaryRouteAnimation != widget.secondaryRouteAnimation
-    || oldWidget.child != widget.child || oldWidget.linearTransition != widget.linearTransition) {
-    _disposeCurve();
-    _setupAnimation();
+    if (
```
Comment (Piinks) [inline packages/flutter/lib/src/cupertino/route.dart:441]: If child changes, why would be not update anymore?

**Review by Piinks — COMMENTED** [review]
@ValentinVignal it looks like there are a lot of changes here unrelated to fixing the leak, some of the changes are identical code, just moved around. I can't actually tell what has changed to fix the leak. Can you revert those unrelated changes?

File: packages/flutter/test/cupertino/route_test.dart (hunk 6 above)
Comment (Piinks) [reply packages/flutter/test/cupertino/route_test.dart:797]: Can you create a new test rather than changing the existing one? We want to preserve the original test to prevent regressions.

File: packages/flutter/lib/src/cupertino/route.dart (hunk 13 above)
Comment (ValentinVignal) [reply packages/flutter/lib/src/cupertino/route.dart:441]: `_setUpAnination()` doesn't use `widget.child` so there is no reason to recompute it when `child` changes.

Moreover, `SizedBox() != SizedBox()` is `true`, so that means that `oldWidget.child != widget.child` will most likely be always true

**Comment by ValentinVignal** [comment]
> @ValentinVignal it looks like there are a lot of changes here unrelated to fixing the leak, some of the changes are identical code, just moved around. I can't actually tell what has changed to fix the leak. Can you revert those unrelated changes?

Sorry about that, it's because we had 2 PRs with conflict. I can try to start everything again

**Comment by ValentinVignal** [comment]
@Piinks @polina-c The changes should have been cleaned now

File: packages/flutter/test/cupertino/route_test.dart (hunk 8 above)
Comment (Piinks) [reply packages/flutter/test/cupertino/route_test.dart:852]: Rather than changing an existing test, can you add a new one? This feel like a potential regression.

File: packages/flutter/lib/src/cupertino/route.dart [hunk 14]
```diff
@@ -435,22 +435,25 @@ class _CupertinoPageTransitionState extends State<CupertinoPageTransition> {
     super.didUpdateWidget(oldWidget);
     if (oldWidget.primaryRouteAnimation != widget.primaryRouteAnimation
     || oldWidget.secondaryRouteAnimation != widget.secondaryRouteAnimation
-    || oldWidget.child != widget.child || oldWidget.linearTransition != widget.linearTransition) {
```
Comment (Piinks) [inline packages/flutter/lib/src/cupertino/route.dart:438]: Is this related to the memory leak? If not, maybe this should be a separate change.

**Review by Piinks — COMMENTED** [review]
Adding @MitchellGoodwin as a reviewer, they have much more context here. :)

File: packages/flutter/test/cupertino/route_test.dart (hunk 8 above)
Comment (MitchellGoodwin) [reply packages/flutter/test/cupertino/route_test.dart:852]: +1 on leaving this test alone as much as possible. It does look like logically the animation wouldn't finish by the time the test tries to go back, so the jump makes sense. I'd rather the test was left as it was to check for regression, even with the jump, and probably renamed. A new test that correctly waits for the animation to finish, then goes back would be preferable.

Does the original test still fail, and how?

File: packages/flutter/lib/src/cupertino/route.dart (hunk 14 above)
Comment (ValentinVignal) [reply packages/flutter/lib/src/cupertino/route.dart:438]: Well now this PR is not about memory leaks anymore since https://github.com/flutter/flutter/pull/147133 got merged in between. Maybe I should rename this PR. Would "Fix jank in cupertino page transition" sounds nice to you ?

File: packages/flutter/test/cupertino/route_test.dart (hunk 8 above)
Comment (ValentinVignal) [reply packages/flutter/test/cupertino/route_test.dart:852]: Those were the minimum changes I found to "let this test alone as much as possible".

What was happening in this test: 
1. Push using the forward animation (500ms)
2. The push is interrupted with a pop at 400ms
3. The pop uses the reversed animation (but should not because it leads to a discontinuity).

After my changes without any modification on the test:

1. Push using the forward animation (500ms)
2. The push is interrupted with a pop at 400ms
3. The pop uses the forward animation for the reverse animation (there is no discontinuity)
4. The test fails at the line
```dart
    await tester.pump(const Duration(milliseconds: 40));
    expect(tester.getTopLeft(find.byType(Placeholder)).dx, moreOrLessEquals(-83.0, epsilon: 1.0));
```

Expected: -83
Actual value: -261

(Comment for reference: https://github.com/flutter/flutter/pull/146999/files#r1575569345)

---

To make this test pass, I have 2 options:

a. I let the push being interrupted and I need to change the expected values to match the use of the forward curve instead of the reverse curved.
b. I let the push animation complete with more `pump`s so the reverse animation for the pop uses the reversed curve and I don't need to change the values in the `expects`.

I have implemented both options. The one linked to the comment is `a.`. And a bit below, there is the `group('Interrupted push', () { ... });` that implements `b.`

---

Is that what you are expecting?

File: packages/flutter/test/cupertino/route_test.dart (hunk 5 above)
Comment (ValentinVignal) [reply packages/flutter/test/cupertino/route_test.dart:1123]: Sure I can do that. I don't mind the `100ms` but I felt it was a bit "hidden"

File: packages/flutter/test/cupertino/route_test.dart [hunk 15]
```diff
@@ -2432,53 +2594,6 @@ void main() {
       expect(tester.getBottomRight(find.byType(Placeholder)).dx, 390.0);
     });
   });
-
-  testWidgets(
-  // TODO(polina-c): remove when fixed https://github.com/flutter/flutter/issues/145600 [leak-tracking-opt-in]
-  experimentalLeakTesting: LeakTesting.settings.withTracked(classes: <String>['CurvedAnimation']),
-  'Fullscreen route does not leak CurveAnimation', (WidgetTester tester) async {
```
Comment (polina-c) [inline packages/flutter/test/cupertino/route_test.dart:2439]: What is the reason to remove this test?

Comment (ValentinVignal) [reply packages/flutter/test/cupertino/route_test.dart:2439]: Oops sorry for that. This must have come from the merge conflicts. I added it back in [test: Add back missing test](https://github.com/flutter/flutter/pull/146999/commits/7a2129a9537bcda7be5fdb28fd36c887ffd84310)

File: packages/flutter/test/cupertino/route_test.dart [hunk 16]
```diff
@@ -2433,7 +2620,7 @@ void main() {
     });
   });
 
-  testWidgets(
+ testWidgets(
```
Comment (polina-c) [inline packages/flutter/test/cupertino/route_test.dart:2623]: extra space?

File: packages/flutter/test/cupertino/route_test.dart (hunk 8 above)
Comment (polina-c) [reply packages/flutter/test/cupertino/route_test.dart:856]: @MitchellGoodwin , does this help?

Comment (MitchellGoodwin) [reply packages/flutter/test/cupertino/route_test.dart:856]: Thank you for the explanation. That all makes sense and the two tests LGTM. Thank you for adjusting the old on and adding a new one.

**Review by polina-c — APPROVED** [review]

**Comment by polina-c** [comment]
@MitchellGoodwin , may I get your formal approval to kick off google testing?

**Review by MitchellGoodwin — APPROVED** [review]
LGTM

**Comment by polina-c** [comment]
Google testing is marked as red, but the tests on CL actually passed. Marking as infra failure.
